sig-and-wg/sig-core/proposals/common-validator.md

```diff
@@ -0,0 +1,266 @@
+
+ci-pr: validate build-image push-image
+ci-master: ci-pr
+ci-release: ci-master
+
+validate: resolve build test format lint
+
+resolve:
+	dep ensure -v -vendor-only
+build:
+	go build -o $(APP_NAME)
+test:
+	go test ./...
+format:
+	go fmt
+lint:
+	go vet
+
+build-image:
+	docker build -t $(APP_NAME) .
+push-image:
+	docker tag $(APP_NAME) $(IMG):$(TAG)
+	docker push $(IMG):$(TAG)
+```
+
+And `makefile` for `ui-api-layer`:
+
+```makefile
+APP_NAME = ui-api-layer
+REPOSITORY_PATH = $(realpath $(shell pwd)/../..)
+include $(REPOSITORY_PATH)/common//template.mk
+```
+
+Thanks to that all targets from `template.mk` are available in `makefile` for `ui-api-layer`.
+
+Unfortunately the structure of components may vary and they may be built in different way. For example `ui-api-layer` has `main.go` file on the root of the component, `binding-usage-controller` has it in `cmd/controller` directory. The template should handle such situation and it is also possible thanks to `template definition` and `eval` function.
+
+After introducing `template definition` the `makefiles` will look like that:
+
+`template.mk`:
+```makefile
+IMG = $(DOCKER_PUSH_REPOSITORY)$(DOCKER_PUSH_DIRECTORY)/$(APP_NAME)
+TAG = $(DOCKER_TAG)
+
+.PHONY: ci-pr ci-master ci-release resolve build test format lint build-image push-image
+
+ci-pr: validate build-image push-image
+ci-master: ci-pr
+ci-release: ci-master
+
+validate: resolve build test format lint
+
+define TARGETS
+resolve:
+	dep ensure -v -vendor-only
+build:
+	go build -o $(APP_NAME) $(1)
+test:
+	go test ./...
+format:
+	go fmt
+lint:
+	go vet
+endef
+
+build-image:
+	docker build -t $(APP_NAME) .
+push-image:
+	docker tag $(APP_NAME) $(IMG):$(TAG)
+	docker push $(IMG):$(TAG)
+```
+
+`ui-api-layer`:
+```makefile
+APP_NAME = ui-api-layer
+REPOSITORY_PATH = $(realpath $(shell pwd)/../..)
+include $(REPOSITORY_PATH)/common//template.mk
+
+$(eval $(call TARGETS,main.go))
+```
+
+Thanks to `template.mk` we will have one place in repository with targets definitions and it will be much easier to maintain it.
+
+>**NOTE**: All components needs to be build by CI system in case of changes in `template.mk`.
+
+### Sandbox Validation
+
+Developers often have different environment than CI system. For example different version of Golang. It is important to provide a way for verifying code against CI environment. For that lets create another target `validate-sandbox`, that will execute all verification in same `buildpack` Docker image as CI system is using.
+
+Definition of `validate-sandbox` target in `template.mk`:
+```makefile
+COMPONENT_REL_PATH=$(shell echo $(shell pwd) | sed 's,$(REPOSITORY_PATH)/,,g')
+
+validate-sandbox:
+	docker run --rm -v "$(REPOSITORY_PATH):/workspace/go/src/github.com/kyma-project/kyma" \
+	--workdir "/workspace/go/src/github.com/kyma-project/kyma" \
+	eu.gcr.io/kyma-project/prow/test-infra/buildpack-golang:v20181119-afd3fbd \
```

Comment:
I have one problem with sandbox version: version of buildpack is duplicated (here and in CI) and may diverge. In such case it will be misleading. Also it assumes all components uses same buildpack which is not necessarily true. Do we care and if so how do we want to solve it?

Reply:
I will add an environment variable to `makefile` on component level with `buildpack` version. Unfortunately, it will be still duplicated in `makefile` and CI job definition. If we will start using new tags for `buildpacks` then it will look better. By new tag, I mean `go1.11` that I introduced with new `buildpack` 😄